### PR TITLE
AP computation for Mask R-CNN output

### PR DESCRIPTION
**Pull request by AlexTS1980**
I re-implemented MS COCO average precision computation for my COVID-19 detection project (segmentation model). The original implementation is matterport. I re-implemented it purely in pytorch.

### CODE REVIEW

**Comment by fmassa** [comment]
Hi,

Thanks for the PR.

Before we move forward, is there a particular reason why we would want to move away from the current AP computation from pycocotools that we currently use in the reference training scripts in https://github.com/pytorch/vision/blob/master/references/detection/coco_eval.py ? They use the official metrics and thus in my opinion are preferable to custom implementations.

**Comment by AlexTS1980** [comment]
> Hi,
> 
> Thanks for the PR.
> 
> Before we move forward, is there a particular reason why we would want to move away from the current AP computation from pycocotools that we currently use in the reference training scripts in https://github.com/pytorch/vision/blob/master/references/detection/coco_eval.py ? They use the official metrics and thus in my opinion are preferable to custom implementations.

Thanks, I stopped the CircleCI build on my account, as requested.

Addressing your issue, to begin with, I didn't know about this wrapper, the torchvision version I've been using (0.3.0) doesn't have  it.  The current one depends on pycocotools, and not everyone has them installed or, needs to use MS COCO, or has the ground truth stored in RLE format. Also, mine is pure pytorch (literally), I don't use numpy, RLE, so I guess the requirements are not so strict, and I don't do any internal conversions. Perhaps it would be good to have both, one specific to MS COCO, the other one for other formats.  

Alex

**Comment by AlexTS1980** [comment]
> Hi,
> 
> Thanks for the PR.
> 
> Before we move forward, is there a particular reason why we would want to move away from the current AP computation from pycocotools that we currently use in the reference training scripts in https://github.com/pytorch/vision/blob/master/references/detection/coco_eval.py ? They use the official metrics and thus in my opinion are preferable to custom implementations.

I tested it on my data(100 images 512x512, https://github.com/AlexTS1980/COVID-CT-Mask-Net/blob/master/evaluate_mean_ap.py), it is very fast (about 2min)

**Comment by fmassa** [comment]
> the torchvision version I've been using (0.3.0) doesn't have it. The current one depends on pycocotools

the [`references/detection` implementation is present since version 0.3.0](https://github.com/pytorch/vision/tree/v0.3.0/references/detection) although it also depended on pycocotools back then.

> Also, mine is pure pytorch (literally), I don't use numpy, RLE, so I guess the requirements are not so strict, and I don't do any internal conversions

Having an implementation in torchvision would be nice, although we would need to have extensive testings to be able to make sure that the implementation is compatible with the mAP implementation in COCO or other sources.

Also, we currently don't have any metrics like mAP in torchvision repo, so adding it would probably involve thinking a bit deeper on how (and if?) we want to expose those. Other metrics such as mean intersection over union (mIoU) are also useful for tasks that we support in torchvision, so this whole discussion deserves a bit more attention.

cc @vfdev-5 for thoughts on adding metrics to torchvision, and how we could structure it.

**Comment by AlexTS1980** [comment]
> > the torchvision version I've been using (0.3.0) doesn't have it. The current one depends on pycocotools
> 
> the [`references/detection` implementation is present since version 0.3.0](https://github.com/pytorch/vision/tree/v0.3.0/references/detection) although it also depended on pycocotools back then.
> 
> > Also, mine is pure pytorch (literally), I don't use numpy, RLE, so I guess the requirements are not so strict, and I don't do any internal conversions
> 
> Having an implementation in torchvision would be nice, although we would need to have extensive testings to be able to make sure that the implementation is compatible with the mAP implementation in COCO or other sources.
> 
> Also, we currently don't have any metrics like mAP in torchvision repo, so adding it would probably involve thinking a bit deeper on how (and if?) we want to expose those. Other metrics such as mean intersection over union (mIoU) are also useful for tasks that we support in torchvision, so this whole discussion deserves a bit more attention.
> 
> cc @vfdev-5 for thoughts on adding metrics to torchvision, and how we could structure it.

I could add a hack that converts pycocotools-specific objects to torch tensors, order of coordinates in boxes specific to MS COCO, etc

**Comment by AlexTS1980** [comment]
Still no reviews. Do I need to add any changes?

**Comment by oke-aditya** [comment]
Hello, metrics implementation for PyTorch is was discussed [here](https://github.com/pytorch/pytorch/issues/22439)

PyTorch Lightning has a metrics package that works well with DP / DDP etc.

**Comment by fmassa** [comment]
@AlexTS1980 standard procedure in PyTorch and torchvision when adding / requesting new features is first to open an issue and discuss about it before sending a PR.

Given that this PR has several consequences to torchvision itself (including if we want to have a metrics package or not within it, given that there are alternatives already, see lightning or ignite for a couple of examples), I would say that this PR is not going to move forward before we decide on an issue beforehand if / how we will be adding metrics to torchvision.

**Comment by AlexTS1980** [comment]
@oke-aditya @fmassa @facebook-github-bot thanks, signed it, thanks. 

As for the existing implementation: I only found the discussion of the related methods, but not a solution. The only one in torch I know of is the one I re-implemented([https://github.com/matterport/Mask_RCNN/blob/master/mrcnn/utils.py#L754](https://github.com/matterport/Mask_RCNN/blob/master/mrcnn/utils.py#L754)). I wrote the script for mask prediction, but I can also easily implement the bounding box too.

**Comment by oke-aditya** [comment]
Hi @AlexTS1980 [this](https://github.com/PyTorchLightning/pytorch-lightning/issues/2552) is an open issue on Lightning and PRs would be welcome (I think) Feel free to Ping people there for mAP metric into lightning.
Lightning has a stable and structured metric package so you might need to reimplement it.

**Comment by AlexTS1980** [comment]
> Hi @AlexTS1980 [this](https://github.com/PyTorchLightning/pytorch-lightning/issues/2552) is an open issue on Lightning and PRs would be welcome (I think) Feel free to open PR there for mAP metric into lightning.
> Lightning has a stable and structured metric package so you might need to reimplement it.

Can I move this PR to Ignite?
